File: pgpainless-core/src/main/java/org/pgpainless/key/generation/KeyRingBuilderInterface.java
// SPDX-FileCopyrightText: 2018 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.key.generation;

import java.security.InvalidAlgorithmParameterException;
import java.security.NoSuchAlgorithmException;
import java.util.Date;
import javax.annotation.Nonnull;

import org.bouncycastle.openpgp.PGPException;
import org.bouncycastle.openpgp.PGPSecretKeyRing;
import org.pgpainless.key.UserId;
import org.pgpainless.s2k.Passphrase;

public interface KeyRingBuilderInterface<B extends KeyRingBuilderInterface<B>> {

    B setPrimaryKey(@Nonnull KeySpec keySpec);

    default B setPrimaryKey(@Nonnull KeySpecBuilder builder) {
        return setPrimaryKey(builder.build());
    }

    B addSubkey(@Nonnull KeySpec keySpec);

    default B addSubkey(@Nonnull KeySpecBuilder builder) {
        return addSubkey(builder.build());
    }

    default B addUserId(UserId userId) {
        return addUserId(userId.toString());
    }

    B addUserId(@Nonnull String userId);

    B addUserId(@Nonnull byte[] userId);

    B setExpirationDate(@Nonnull Date expirationDate);

    B setPassphrase(@Nonnull Passphrase passphrase);

    PGPSecretKeyRing build() throws NoSuchAlgorithmException, PGPException,
                InvalidAlgorithmParameterException;
}

File: pgpainless-core/src/main/kotlin/org/pgpainless/key/generation/KeyRingBuilderInterface.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.key.generation

import java.security.InvalidAlgorithmParameterException
import java.security.NoSuchAlgorithmException
import java.util.*
import org.bouncycastle.openpgp.PGPException
import org.bouncycastle.openpgp.PGPSecretKeyRing
import org.pgpainless.util.Passphrase

interface KeyRingBuilderInterface<B : KeyRingBuilderInterface<B>> {

    fun setPrimaryKey(keySpec: KeySpec): B

    fun setPrimaryKey(builder: KeySpecBuilder): B = setPrimaryKey(builder.build())

    fun addSubkey(keySpec: KeySpec): B

    fun addSubkey(builder: KeySpecBuilder): B = addSubkey(builder.build())

    fun addUserId(userId: CharSequence): B

    fun addUserId(userId: ByteArray): B

    fun setExpirationDate(expirationDate: Date?): B

    fun setPassphrase(passphrase: Passphrase): B

    @Throws(
        NoSuchAlgorithmException::class,
        PGPException::class,
        InvalidAlgorithmParameterException::class)
    fun build(): PGPSecretKeyRing
}
